crypto/params.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+
+#define SET_RETURN_SIZE(p, sz) \
+    if (p->return_size != NULL) \
+        *p->return_size = (sz)
+
+static const OSSL_PARAM *param_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+#define PARAM_INT_GET(to, from) \
+    case sizeof(from): \
+        if (sizeof(to) >= sizeof(from)) { \
+            *val = (to)*(const from *)p->buffer; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT_SET(to, from) \
+    case sizeof(to): \
+        SET_RETURN_SIZE(p, sizeof(to)); \
+        if (sizeof(to) >= sizeof(from)) { \
+            *(to *)p->buffer = (to)val; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT(name, type, pre) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, const char *key, type *val) \
+    { \
+        if (val == NULL \
+            || (p = param_locate(p, key)) == NULL \
```

Comment:
Oooooh!  I just thought of something.  What if the caller has set up this:

``` C
    int foo_foo = 42;
    const OSSL_PARAM foo[] = {
        OSSL_PARAM_INT("foo", foo_foo),
        OSSL_PARAM_END
    };
```

BUT, the provider actually provides two parameters, "foo" and "bar".  So I expect that the provider code will be something like this:

``` C
    if (!OSSL_PARAM_get_int(params, "foo", &obj->foo)
        || !OSSL_PARAM_get_int(params, "bar", &obj->bar))
        return 0;  /* #fail */
```

However, by definition, *not providing exactly all parameters is not an error per se*...  It's up to the provider to decide what parameters need to be set together.

This is even more important for `OSSL_PARAM_set_##name`.  Why should the caller set up a `OSSL_PARAM` array that contains "bar" if the caller is only interested in "foo"?

Therefore, `param_locate(p, key) == NULL` cannot be an error

Reply:
I don't see how `param_locate(p, key) == NULL` could be considered a success though.

I could add a _default value_ for the get calls which is returned for non-fatal problems.

The definition of non-fatal being deliberately omitted for the moment but param_locate() == NULL would be such.

Reply:
One side need not know that the other is interested in _bar_.  I think it should be up to the side that wants _bar_ to deal with its absence.

Reply:
So a provider would have to try and locate the parameter it wants deal before calling one of these functions *which will try and locate them again*?

Reply:
I'm thinking that if it's up to the caller to decide what to do in case of "missing" parameters, then the caller should call `OSSL_PARAM_locate`, and the rest if your code should only deal with the OSSL_PARAM pointer they receive (i.e. as a single item) and not do the locate magic.

Reply:
This is an easy enough change.

Let's see which flavour is chosen before I rip the locates out -- both have them at present I believe.